pandas/core/groupby/ops.py

```diff
@@ -671,12 +670,7 @@ def _cython_operation(
         elif is_bool_dtype(dtype):
             values = ensure_int_or_float(values)
         elif is_integer_dtype(dtype):
-            # we use iNaT for the missing value on ints
-            # so pre-convert to guard this condition
-            if (values == iNaT).any():
-                values = ensure_float64(values)
-            else:
-                values = ensure_int_or_float(values)
```

Comment:
I think these iNaT checks are necessary because in algos like `group_min_max`, it assumes that it is impossible for an integer array to have iNaT (because it assumes `datetimelike=True` and safety in setting missing values as `NPY_NAT`).

eg on this branch:
```
ser = pd.Series([1, iNaT])
print(ser.groupby([1, 1]).max(min_count=2))
```
gives
```
1   -9223372036854775808
dtype: int64
```
because the iNaT is interpreted as NaN, so `min_count` isn't reached. (Then the NaN is set with iNaT, but not interpreted as NaN anymore)

Forcing mask usage (https://github.com/pandas-dev/pandas/pull/40651#discussion_r606297863) would be another way to handle this more robustly (since to specify missing values, the mask itself can just be modified inplace, precluding the need for the existing iNaT workarounds to signify missing values in the algo result)

Reply:
> because it assumes datetimelike=True

isn't this the underlying problem in this example?

Reply:
I think the reason datetimelike isn't used in `group_min_max`(and some other groupby algos) is because of the additional problem that there is no way to encode a missing value in an integer (not datetimelike) array. 

So the current logic casts integer to float if iNaT is present, so that iNaT can be used to signify missing values for integer data (regardless of datetimelike status).

Reply:
> that there is no way to encode a missing value in an integer (not datetimelike) array.

i think we dont _need_ to encode it because we have `counts==0` (or counts<min_count) to convey that information

Reply:
if we use masks we could avoid this awkwardness but that will need some refactoring

Reply:
its not too hard to edit group_min_max to handle this example (will push in a bit).  We can improve matters a bit by using `counts < min_count` as a mask more consistently, but that won't do anything about int64->float64 conversions being lossy.  a full solution to that would be one of a) explicitly be OK with the lossiness, b) detect when the conversion would be lossy and cast to either object or float128, c) return an IntegerArray